libnmstate/nm/nmclient.py

```diff
@@ -220,3 +229,57 @@ def _register_first_action(self):
     def _execute_action_once(self, _):
         self.execute_next_action()
         return False
+
+
+class _NmClient:
+    def __init__(self):
+        self._client = NM.Client.new(None)
+        logging.debug("NM.Client created")
+        self._mainloop = None
+
+    @property
+    def client(self):
+        return self._client
+
+    def delete_client(self):
+        if not self.client:
+            return
+        try:
+            ref_count = sys.getrefcount(self.client)
+            if ref_count != 2:
+                # The getrefcount() hold 1 reference and self.client itself
+                # hold 1 reference.
+                raise error.NmstateInternalError(
+                    "_NmClient.delete_client() cannot release resources "
+                    "of NM.Client duo to other unreleased use."
+                )
+            if NM.Client.get_version(self.client) >= StrictVersion("1.21.0"):
+                self._nmclient_cleanup()
+            else:
+                self._nmclient_cleanup_old()
+        except Exception:
+            pass
+
+    def __del__(self):
+        self.delete_client()
```

Comment:
I would prefer to raise here in case the reference is incorrect (there is a leak).
The code should explicitly cleanup, if at the gc cleanup stage it is not clean, there is a bug in our code and we should blow up.

Doing things in the destructor is dangerous, because it is unpredictable when it occurs (what happens before or after). Usually only simple small things are done at this stage, something that is limited in scope.

But if you feel strongly about this, we can delay this discussion to a following PR.